lib/compress/zstd_opt.c

```diff
@@ -733,6 +735,61 @@ U32 ZSTD_insertBtAndGetAllMatches (
         }
     }
 
+    /* Handle long distance matches if applicable. */
+    if (mnum > 0 && ms->ldmSeqStore && ms->ldmSeqStore->size > 0) {
+        rawSeqStore_t* ldmSeqStore = ms->ldmSeqStore;
+        int ldmIndex = ZSTD_ldm_hasMatchAtAbsolutePosition(ldmSeqStore, curr);
+        if (ldmIndex >= 0) {
+            rawSeq possibleLdm = ldmSeqStore->seq[ldmIndex];
+            U32 matchLength = possibleLdm.matchLength;
+            U32 offset = possibleLdm.offset;
+
+            /* needed for ZSTD_ldm_maybeSplitSequence() to split the correct seq */
+            ldmSeqStore->pos = ldmIndex;
+            ldmSeqStore->bytesRead = curr - 1;
+
+            /* TODO: as followup, also insert matches that are shorter than bestLength, but with smaller offset */
+            if (matchLength >= bestLength) {
+                /* TODO: This calculation needs more scrutiny, but seems to work to split matchLengths
+                 * that are longer than the remaining bytes in a block. But this doesn't work for litLengths
+                 * longer than a block...
+                 * */
+                U32 remainingBytes = (U32)(iLimit - ip + ldmSeqStore->seq[ldmIndex].litLength);
+                /* rawSeqStore generated by multithreaded compression may result in match lengths longer than a block.
+                 * So there is a possibility that we must split the sequence across block boundaries.
+                 * ZSTD_ldm_maybeSplitSequence() mutates the sequence at "pos" and returns a valid sequence.
+                 */
+                rawSeq finalSeq = ZSTD_ldm_maybeSplitSequence(ldmSeqStore, remainingBytes, minMatch);
+                if (ldmSeqStore->pos != ldmIndex) {
+                    /* no sequence split */
+                    ldmSeqStore->bytesRead += finalSeq.matchLength;
+                } else {
+                    if (finalSeq.offset == 0) {
+                        /* This is the case that I'm not really sure how to handle yet
+                         * Offset == 0 implies that the litLength > remaining bytes
```

Comment:
> Offset == 0 implies that the litLength > remaining bytes

I don't understand this sentence.
Both parts of the statement don't align naturally together.
Is there something in the code that makes it happen ?